File: examples/connext_dds/distributed_logger/py/dist_logger_example.py
#
# (c) 2024 Copyright, Real-Time Innovations, Inc.  All rights reserved.
#
# RTI grants Licensee a license to use, modify, compile, and create derivative
# works of the Software solely for use with RTI products.  The Software is
# provided "as is", with no warranty of any type, including any warranty for
# fitness for any purpose. RTI is under no obligation to maintain or support
# the Software.  RTI shall not be liable for any incidental or consequential
# damages arising out of the use or inability to use the software.
#

import time
import argparse
import rti.connextdds as dds
import rti.logging.distlog as distlog

def distlogger_example(application_kind, domain_id, sleep, iterations):
    # First, create the options to personalize Distributed Logger.
    # If no options are provided, default ones will be created.
    options = distlog.LoggerOptions()
    options.domain_id = domain_id
    options.application_kind = application_kind

    # Then, set the created options. The method init must be called before
    # using the Logger instance.
    distlog.Logger.init(options)

    # The log messages are published as DDS topics, which allows your DDS
    # applications to subscribe to them. You can also run rtiddsspy or
    # RTI Admin Console to visualize the logs.
    for i in range(1, iterations + 1):
        print(f"\nIteration #{i}")

        distlog.Logger.debug("This is a debug message")
        distlog.Logger.warning("This is a warning message")
        distlog.Logger.error("This is an error message")

        time.sleep(sleep)

    distlog.Logger.finalize()


def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d", "--domain_id", type=int, default=0, help="Domain ID (default: 0)"
    )
    parser.add_argument(
        "-s",
        "--sleep",
        type=int,
        default=1,
        help="Seconds between iterations (default: 1)",
    )
    parser.add_argument(
        "-i",
        "--iterations",
        type=int,
        default=50,
        help="Number of iterations (default: 50)",
    )
    args = parser.parse_args()

    try:
        distlogger_example(
            "Python Distributed Logger Example",
            args.domain_id,
            args.sleep,
            args.iterations,
        )
    except Exception as ex:
        print(f"Exception in distlogger_example: {ex}")

    dds.DomainParticipant.finalize_participant_factory()


if __name__ == "__main__":
    main()

File: examples/connext_dds/distributed_logger/c++11/DistLoggerExample.cxx
//
// (c) 2023 Copyright, Real-Time Innovations, Inc.  All rights reserved.
//
//  RTI grants Licensee a license to use, modify, compile, and create derivative
//  works of the Software.  Licensee has the right to distribute object form
//  only for use with RTI products.  The Software is provided "as is", with no
//  warranty of any type, including any warranty for fitness for any purpose.
//  RTI is under no obligation to maintain or support the Software.  RTI shall
//  not be liable for any incidental or consequential damages arising out of the
//  use or inability to use the software.
//

#include <iostream>

#include <rti/util/util.hpp>  // for rti::util::sleep()
#include <rti/distlogger/DistLogger.hpp>

using namespace std;
using namespace dds::domain;
using namespace rti::dist_logger;


void distlogger_example_main(
        const string application_kind,
        const int domain_id,
        const int sleep,
        const uint iterations)
{
    // First, create the options to personalize Distributed Logger.
    // If no options are provided, default ones will be created.
    DistLoggerOptions options;
    options.domain_id(domain_id);
    options.application_kind(application_kind);

    // Then, set the created options.
    // You can only call set_options before getting the Distributed Logger
    // instance. Once an instance has been created, attempting to call
    // set_options will throw an exception.
    DistLogger::set_options(options);

    // Instantiate Distributed Logger
    DistLogger dist_logger = DistLogger::get_instance();


    // The log messages are published as DDS topics, which allows your DDS
    // applications to subscribe to them. You can also run rtiddsspy or
    // RTI Admin Console to visualize the logs.
    for (uint i = 1; i <= iterations; ++i) {
        cout << "\nIteration #" << i << endl;

        dist_logger.debug("This is a debug message");
        dist_logger.warning("This is a warning message");
        dist_logger.error("This is an error message");

        rti::util::sleep(dds::core::Duration(sleep));
    }

    // The DistLogger instance must be finalized for clean-up
    // before the the participant factory is finalized.
    DistLogger::finalize();
}


int main(int argc, char *argv[])
{
    const string application = "Modern C++ Distributed Logger Example";

    int domain_id = 0;
    int sleep = 1;
    uint iterations = 50;

    for (int i = 1; i < argc;) {
        const string &param = argv[i++];

        if ((param == "-d" || param == "--domain_id") && i < argc) {
            domain_id = atoi(argv[i++]);
        } else if ((param == "-i" || param == "--iterations") && i < argc) {
            iterations = atoi(argv[i++]);
        } else if ((param == "-s" || param == "--sleep") && i < argc) {
            sleep = atoi(argv[i++]);
        } else {
            cout << argv[0] << " [options]\n"
                 << "\t-d, --domain_id <domain ID> (default: 0)\n"
                 << "\t-s, --sleep <seconds between iterations> (default: 1)\n"
                 << "\t-i, --iterations <number of iterations> (default: 50)\n"
                 << "\t-h, --help Display this message.\n"
                 << endl;
            return -1;
        }
    }

    try {
        distlogger_example_main(application, domain_id, sleep, iterations);

        // Finalize Domain Participant Factory
        dds::domain::DomainParticipant::finalize_participant_factory();
    } catch (const exception &ex) {
        // This will catch DDS exceptions
        cerr << "Exception in distlogger_example_main: " << ex.what() << endl;
        return -1;
    }

    return 0;
}
